presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java

```diff
@@ -348,23 +349,23 @@ private static String formatFragment(FunctionRegistry functionRegistry, StatsCal
         builder.append(indentString(1)).append(format("Execution Flow: %s\n", fragment.getPipelineExecutionStrategy()));
 
         if (stageInfo.isPresent()) {
-            builder.append(textLogicalPlan(fragment.getRoot(), fragment.getSymbols(), functionRegistry, statsCalculator, costCalculator, session, planNodeStats.get(), 1, verbose))
+            builder.append(textLogicalPlan(fragment.getRoot(), TypeProvider.copyOf(fragment.getSymbols()), functionRegistry, statsCalculator, costCalculator, session, planNodeStats.get(), 1, verbose))
```

Comment:
should not `fragment` also return `TypeProvider`? Then `getSymbols` should be renamed to `getTypes`.

Reply:
No, fragment needs to be json-serializable. I didn't want to go through the trouble of making TypeProvider work that way. Keep in mind that TypeProvider is just so that we can split the AST from plan IR, add type information to IR nodes and fix all usages gradually. I expect to get rid of it when I'm done with that.